engine/execution/storehouse/in_memory_register_store.go

```diff
@@ -0,0 +1,336 @@
+	parentByBlockID    map[flow.Identifier]flow.Identifier                        // for register updates to be fork-aware
+	blockIDsByHeight   map[uint64]map[flow.Identifier]struct{}                    // for pruning
+	prunedHeight       uint64                                                     // registers at pruned height are pruned (not saved in registersByBlockID)
+	prunedID           flow.Identifier                                            // to ensure all blocks are extending from pruned block (last finalized and executed block)
+}
+
+func NewInMemoryRegisterStore(lastHeight uint64, lastID flow.Identifier) *InMemoryRegisterStore {
+	return &InMemoryRegisterStore{
+		registersByBlockID: make(map[flow.Identifier]map[flow.RegisterID]flow.RegisterValue),
+		parentByBlockID:    make(map[flow.Identifier]flow.Identifier),
+		blockIDsByHeight:   make(map[uint64]map[flow.Identifier]struct{}),
+		prunedHeight:       lastHeight,
+		prunedID:           lastID,
+	}
+}
+
+// SaveRegisters saves the registers of a block to InMemoryRegisterStore
+// It needs to ensure the block is above the pruned height and is connected to the pruned block
+func (s *InMemoryRegisterStore) SaveRegisters(
+	height uint64,
+	blockID flow.Identifier,
+	parentID flow.Identifier,
+	registers []flow.RegisterEntry,
+) error {
+	// preprocess data before acquiring the lock
+	regs := make(map[flow.RegisterID]flow.RegisterValue, len(registers))
+	for _, reg := range registers {
+		regs[reg.Key] = reg.Value
+	}
+
+	s.Lock()
+	defer s.Unlock()
+
+	// ensure all saved registers are above the pruned height
+	if height <= s.prunedHeight {
+		return fmt.Errorf("saving pruned registers height %v <= pruned height %v", height, s.prunedHeight)
+	}
+
+	// ensure the block is not already saved
+	_, ok := s.registersByBlockID[blockID]
+	if ok {
+		// already exist
+		return fmt.Errorf("saving registers for block %s, but it already exists", blockID)
+	}
+
+	// make sure parent is a known block or the pruned block, which forms a fork
+	_, ok = s.registersByBlockID[parentID]
+	if !ok {
+		// parent doesn't exist, check if it is the pruned block
+		if parentID != s.prunedID {
+			return fmt.Errorf("saving registers for block %s, but its parent %s is not saved", blockID, parentID)
+		}
+	}
+
+	// update registers for the block
+	s.registersByBlockID[blockID] = regs
+
+	// update index on parent
+	s.parentByBlockID[blockID] = parentID
+
+	// update index on height
+	sameHeight, ok := s.blockIDsByHeight[height]
+	if !ok {
+		sameHeight = make(map[flow.Identifier]struct{})
+		s.blockIDsByHeight[height] = sameHeight
+	}
+
+	sameHeight[blockID] = struct{}{}
+	return nil
+}
+
+// GetRegister will return the latest updated value of the given register
+// since the pruned height.
+// It returns ErrPruned if the register is unknown or not updated since the pruned height
+// Can't return ErrNotFound, since we can't distinguish between not found or not updated since the pruned height
```

Comment:
is that true? what about the case where we want to get a register above the prune height but it's not found?

Reply:
but what about the case where you have both errors, you still keep the ErrPruned but only in the case where the height is above the pruned height and register is not found you return not found error, I believe in that case the pruned height is not relevant to the consumer of the API since it just tells them the register doesn't exist.

Reply:
A `storage.NotFound` error to me means the register does not exist at a certain height. Rather than not found in in-memory register store, because we are returning `storage.NotFound` instead of `inMemoryStore.NotFound`. If a register is not found in in-memory-register, in only means it's not updated within the range between the pruned height to the given height, however, it might exist and updated in a block below the pruned height. That's why I prefer to return PrunedError, instead of storage.NotFound. 

So, let's say in memory register store's pruned height is 20, and RegisterA is updated at height 10, which means the in-memory register store does not have its value. Then if I get register A at height 30, It's better to return PrunedError with prunedHeight, instead of storage.NotFound without prunedHeight, [so that RegisterStore can just query OnDiskRegsiterStore to check whether it actually exists or not. And RegisterStore will convert the PrunedError to NotFound error](https://github.com/onflow/flow-go/pull/4940/files#diff-23aa2b9b91f747982e68750d897cc867ad7c04632610949d83b1f284724f92d6R81). In other words, user of RegisterStore never see `PrunedError`. `PrunedError` is only used between InMemoryRegisterStore and RegisterStore.